autopush-common/src/db/bigtable/bigtable_client/mod.rs

```diff
@@ -650,6 +651,64 @@ impl DbClient for BigTableClientImpl {
         self.write_row(row).await.map_err(|e| e.into())
     }
 
+    /// Add channels in bulk (used mostly during migration)
+    ///
+    async fn add_channels(&self, uaid: &Uuid, channels: HashSet<Uuid>) -> DbResult<()> {
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .map_err(|e| DbError::General(e.to_string()))?
+            .as_millis();
+        let mut entries = protobuf::RepeatedField::default();
+        let mut req = bigtable::MutateRowsRequest::default();
+        let mut limit: u32 = 0;
+        req.set_table_name(self.settings.table_name.clone());
+
+        // Create entries that define rows that contain mutations to hold the updated value which
+        // will create/update the channels.
+        for channel in channels {
+            let mut entry = bigtable::MutateRowsRequest_Entry::default();
+            let key = as_key(uaid, Some(&channel), None);
+            entry.set_row_key(key.into_bytes());
+
+            let mut cell_mutations = protobuf::RepeatedField::default();
+            let mut mutation = data::Mutation::default();
+            let mut set_cell = data::Mutation_SetCell {
+                family_name: ROUTER_FAMILY.to_owned(),
+                ..Default::default()
+            };
+            set_cell.set_column_qualifier("updated".to_owned().into_bytes().to_vec());
+            set_cell.set_value(now.to_be_bytes().to_vec());
+
+            mutation.set_set_cell(set_cell);
+            cell_mutations.push(mutation);
+            entry.set_mutations(cell_mutations);
+            entries.push(entry);
+            // There is a limit of 100,000 mutations per batch for bigtable.
+            // https://cloud.google.com/bigtable/quotas
+            // If you have 100,000 channels, you have too many.
+            limit += 1;
+            if limit >= 100_000 {
+                break;
+            }
+        }
+        req.set_entries(entries);
+
+        let bigtable = self.pool.get().await?;
+
+        // ClientSStreamReceiver will cancel an operation if it's dropped before it's done.
+        let resp = bigtable
+            .conn
+            .mutate_rows(&req)
+            .map_err(error::BigTableError::Write)?;
+        // TODO: is there a better way to do this? I don't think we care about the return stream.
```

Comment:
A problem is that it's a stream of results of each individual mutation. It appears some in the batch can succeed while others fail. I think at the very least we need to loop through the stream to check every result: this currently only checks the first result.